Lib/asyncio/coroutines.py

```diff
@@ -107,12 +107,14 @@ def coroutine(func):
     If the coroutine is not yielded from before it is destroyed,
     an error message is logged.
     """
-    if inspect.iscoroutinefunction(func):
+    if (inspect.iscoroutinefunction(func)
+            and not isinstance(func, functools.partial)):
         # In Python 3.5 that's all we need to do for coroutines
         # defined with "async def".
         return func
 
-    if inspect.isgeneratorfunction(func):
+    if (inspect.isgeneratorfunction(func)
+            and not isinstance (func, functools.partial)):
```

Comment:
We don't need this extra check either -- if `func` is a generator function, it means that we don't need to wrap it into a proxy to force it to return a generator.